syntax = "proto3";

package task_manager;

option go_package = "task-manager-service/pkg/proto";

// 任务管理服务
service TaskManagerService {
    // 提交任务
    rpc SubmitTask(TaskSubmitRequest) returns (TaskSubmitResponse);
    
    // 查询任务状态
    rpc GetTaskStatus(TaskStatusRequest) returns (TaskStatusResponse);
    
    // 批量提交任务
    rpc SubmitBatchTasks(BatchTaskSubmitRequest) returns (BatchTaskSubmitResponse);
    
    // 任务状态流式监听
    rpc WatchTaskStatus(TaskWatchRequest) returns (stream TaskStatusUpdate);
    
    // 取消任务
    rpc CancelTask(TaskCancelRequest) returns (TaskCancelResponse);
    
    // 获取任务列表
    rpc ListTasks(TaskListRequest) returns (TaskListResponse);
}

// 任务提交请求
message TaskSubmitRequest {
    string task_type = 1;           // 任务类型 (document_processing, embedding_generation, vector_storage)
    string service_name = 2;        // 来源服务名
    string knowledge_base_id = 3;   // 知识库ID
    string priority = 4;            // 任务优先级 (high, medium, low)
    map<string, string> payload = 5; // 任务载荷
    int32 max_retries = 6;          // 最大重试次数
    int32 timeout_seconds = 7;      // 超时时间
    string callback_url = 8;        // 回调URL
}

// 任务提交响应
message TaskSubmitResponse {
    string task_id = 1;             // 任务ID
    string status = 2;              // 任务状态
    string message = 3;             // 响应消息
    int64 created_at = 4;           // 创建时间戳
    int64 estimated_completion = 5; // 预计完成时间
    string queue_name = 6;          // 队列名称
}

// 任务状态查询请求
message TaskStatusRequest {
    string task_id = 1;             // 任务ID
}

// 任务状态响应
message TaskStatusResponse {
    string task_id = 1;             // 任务ID
    string status = 2;              // 状态 (pending, processing, completed, failed, cancelled)
    int32 progress = 3;             // 进度百分比
    string message = 4;             // 状态消息
    string error_message = 5;       // 错误信息
    int64 created_at = 6;           // 创建时间戳
    int64 started_at = 7;           // 开始时间戳
    int64 completed_at = 8;         // 完成时间戳
    map<string, string> metadata = 9; // 任务元数据
    string worker_id = 10;          // 工作节点ID
}

// 批量任务提交请求
message BatchTaskSubmitRequest {
    string batch_id = 1;            // 批次ID
    repeated TaskSubmitRequest tasks = 2; // 任务列表
    bool wait_for_all = 3;          // 是否等待所有任务完成
}

// 批量任务提交响应
message BatchTaskSubmitResponse {
    string batch_id = 1;            // 批次ID
    repeated TaskSubmitResponse responses = 2; // 响应列表
    int32 total_count = 3;          // 总任务数
    int32 submitted_count = 4;      // 已提交数
    int32 failed_count = 5;         // 提交失败数
}

// 任务监听请求
message TaskWatchRequest {
    string task_id = 1;             // 任务ID
    string batch_id = 2;            // 批次ID (可选)
    bool include_logs = 3;          // 是否包含日志
}

// 任务状态更新
message TaskStatusUpdate {
    string task_id = 1;             // 任务ID
    string status = 2;              // 最新状态
    int32 progress = 3;             // 进度百分比
    string message = 4;             // 状态消息
    int64 timestamp = 5;            // 时间戳
    map<string, string> metadata = 6; // 元数据
    repeated string logs = 7;       // 日志信息
}

// 任务取消请求
message TaskCancelRequest {
    string task_id = 1;             // 任务ID
    string reason = 2;              // 取消原因
}

// 任务取消响应
message TaskCancelResponse {
    string task_id = 1;             // 任务ID
    bool success = 2;               // 是否成功
    string message = 3;             // 响应消息
    int64 cancelled_at = 4;         // 取消时间戳
}

// 任务列表查询请求
message TaskListRequest {
    repeated string task_types = 1;  // 任务类型过滤
    repeated string statuses = 2;    // 状态过滤
    string knowledge_base_id = 3;    // 知识库ID过滤
    int32 limit = 4;                // 限制数量
    int32 offset = 5;               // 偏移量
    string sort_by = 6;             // 排序字段
    string sort_order = 7;          // 排序顺序 (asc, desc)
}

// 任务列表响应
message TaskListResponse {
    repeated TaskStatusResponse tasks = 1; // 任务列表
    int32 total_count = 2;          // 总数
    int32 limit = 3;                // 限制数量
    int32 offset = 4;               // 偏移量
    bool has_more = 5;              // 是否有更多
}